products/cloudiot/api.yaml

```diff
@@ -0,0 +1,194 @@
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: region
+        input: true
+        url_param_only: true
+        required: true
+        description: |
+          The region of this Device Registry.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'id'
+        input: true
+        required: true
+        description: |
+          The unique identifier for the device registry. For example,
+          `myRegistry`.
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource path name. For example,
+          `projects/example-proj/locations/us-central1/registries/my-registry`.
+      - !ruby/object:Api::Type::Array
+        name: 'eventNotificationConfigs'
+        description: |
+          List of configurations for event notifications, such as PubSub topics
+          to publish device events to.
+        max_size: 10
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::String
+              name: 'subfolderMatches'
+              description: |
+                If the subfolder name matches this string exactly, this
+                configuration will be used. The string must not include the
+                leading '/' character. If empty, all strings are matched. Empty
+                value can only be used for the last `event_notification_configs`
+                item.
+            - !ruby/object:Api::Type::String
+              name: 'pubsubTopicName'
+              required: true
+              description: |
+                PubSub topic name to publish device events.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'stateNotificationConfig'
+        description: |
+          A PubSub topic to publish device state updates.
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'pubsubTopicName'
+            required: true
+            description: |
+              PubSub topic name to publish device state updates.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'mqttConfig'
+        description: |
+          Activate or deactivate MQTT.
+        properties:
+          - !ruby/object:Api::Type::Enum
+            name: 'mqttEnabledState'
+            description: |
+              The field allows `MQTT_ENABLED` or `MQTT_DISABLED`
+            required: true
+            values:
+              - :MQTT_ENABLED
+              - :MQTT_DISABLED
+      - !ruby/object:Api::Type::NestedObject
+        name: 'httpConfig'
+        description: |
+          Activate or deactivate HTTP.
+        properties:
+          - !ruby/object:Api::Type::Enum
+            name: 'httpEnabledState'
+            required: true
+            description: |
+              The field allows `HTTP_ENABLED` or `HTTP_DISABLED`.
+            values:
+              - :HTTP_ENABLED
+              - :HTTP_DISABLED
+      - !ruby/object:Api::Type::Enum
```

Comment:
Can you add `send_empty_value: true` to this field to line up with https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_cloudiot_registry.go#L250?

Reply:
done

Reply:
If I add `send_empty_value: true` to `logLevel`, my `TestAccCloudIoTRegistry_update` breaks, because it sends an empty `log_level` in the HTTP request, and the API doesn't like that (that field is an Enum that doesn't accept empty strings).

Should I remove that field from the updateMask if it's set to an empty string (or if the user doesn't set it)? Or am I missing something?

thanks!

```
---[ REQUEST ]---------------------------------------
PATCH /v1/projects/xxxxxxxxxxxxx/locations/us-central1/registries/psregistry-test-4d3gub0mql?alt=json&updateMask=eventNotificationConfigs%2Cstate_notification_config.pubsub_topic_name%2Cmqtt_config.mqtt_enabled_state%2Chttp_config.http_enabled_state%2Ccredentials HTTP/1.1
Host: cloudiot.googleapis.com
User-Agent: HashiCorp Terraform/0.12.7-sdk (+https://www.terraform.io) Terraform Plugin SDK/1.4.0 terraform-provider-google/dev
Content-Length: 1704
Content-Type: application/json
Accept-Encoding: gzip

{
 "credentials": [
  {
   "publicKeyCertificate": {
    "certificate": "-----BEGIN CERTIFICATE-----\naAAAAAAAAAA\n-----END CERTIFICATE-----\n",
    "format": "X509_CERTIFICATE_PEM"
   }
  }
 ],
 "eventNotificationConfigs": [
  {
   "pubsubTopicName": "projects/xxxxxxxxxxxxxxxxx/topics/psregistry-additional-test-telemetry-psregistry-additional-test-telemetry-iinb3u43lq",
   "subfolderMatches": "test/directory"
  },
  {
   "pubsubTopicName": "projects/xxxxxxxxxxxxxxx/topics/psregistry-test-telemetry-psregistry-test-telemetry-2o11c0orz4"
  }
 ],
 "httpConfig": {
  "httpEnabledState": "HTTP_DISABLED"
 },
 "logLevel": "",
 "mqttConfig": {
  "mqttEnabledState": "MQTT_DISABLED"
 },
 "stateNotificationConfig": {
  "pubsubTopicName": "projects/xxxxxxxxx/topics/psregistry-test-devicestatus-psregistry-test-devicestatus-kxqivokfwo"
 }
}
```

```
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Alt-Svc: xxxxxx
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Thu, 07 May 2020 22:30:34 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

1cb
{
  "error": {
    "code": 400,
    "message": "Invalid value at 'device_registry.log_level' (TYPE_ENUM), \"\"",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "device_registry.log_level",
            "description": "Invalid value at 'device_registry.log_level' (TYPE_ENUM), \"\""
          }
        ]
      }
    ]
  }
}
```

Reply:
Here's what I think is going on! The handwritten version of this resource has a few typos in it that nobody ever caught:
* https://github.com/terraform-providers/terraform-provider-google/blob/93047df43006806540f4fe6c013ccf5825be370e/google/resource_cloudiot_registry.go#L250 should be LogLevel
* https://github.com/terraform-providers/terraform-provider-google/blob/93047df43006806540f4fe6c013ccf5825be370e/google/resource_cloudiot_registry.go#L333 should be logLevel

So it looked like we were sending the empty value, but we actually weren't. It also looked like we were sending a valid updateMask, but we weren't. My guess is that if the update mask isn't valid, the API ignores it and treats it like a PUT (I'm not quite sure why, but that does seem to be what's happening).

This means that sending an empty log level would have accidentally worked (because the API doesn't send a log level back if it wasn't sent in the request), and updating a log level to empty would also have accidentally worked (because the API was replacing the whole resource with what we sent).

What I think we should actually do is set a default value of "NONE", which I believe from looking at the docs is what the API does. Then I think we can remove the DiffSuppressFunc, but we may need to replace it with `emptyOrDefaultStringSuppress("NONE")`. I'll let you figure that out. And then don't set `send_empty_value`.

This also made me realize that some of the updateMask values in your handwritten code were written as snake_case, but should be camelCase.

Give that all a try and let me know how it goes!

Reply:
Thanks for figuring this out! I'm working on this change now. Happy to see that we're going to fix a bug as well!

Reply:
@danawillow I've pushed a couple of commits with the requested changes:

1. Set the default value for log level to `:NONE` in `api.yaml`.
1. Don't set `send_empty_value` for log level.
1. Change `emptyOrDefaultStringSuppress("")` to `emptyOrDefaultStringSuppress("NONE")`.
1. Fix case of some values when populating the update mask.

All my test cases pass!

Reply:
(and Cloud Build is happy too!)